dep0/lib/03_typecheck/src/beta_reduction.cpp

```diff
@@ -175,12 +178,14 @@ bool beta_normalize(body_t& body)
         changed |= impl::beta_normalize(*it);
         it = match(
             it->value,
-            [&] (expr_t::app_t const&)
+            [&] (expr_t::app_t const& app)
             {
-                // currently all functions are immutable and can be dropped right away;
-                // eventually we'll have to keep calls to mutable functions,
-                // in order to retain their side effects
-                return body.stmts.erase(it);
+                // calls to immutable functions can be removed since they do nothing; keep anything else
```

Comment:
Still have to keep arguments which are themselves calls to mutable functions, and so on recursively